File: kmp.py
def constructLPS(pat, lps):
    len_, i, m = 0, 1, len(pat)

    while i < m:
        if pat[i] == pat[len_]:
            len_ += 1
            lps[i] = len_
            i += 1
        else:
            if len_ != 0:
                len_ = lps[len_-1]
            else:
                lps[i] = 0
                i += 1


def kmp(pat, txt, lps):
    i, j, m, n = 0, 0, len(txt), len(pat)

    while i < m:
        if txt[i] == pat[j]:
            i += 1
            j += 1
            if j == n:
                print("Pattern found at "+str(i-n))
                j = lps[j-1]
        else:
            if j == 0:
                i += 1
            else:
                j = lps[j-1]


txt = "aabaacaadaabaaba"
pat = "aaba"
lps = [0]*len(pat)
constructLPS(pat, lps)

kmp(pat, txt, lps)


# Time Complexity: O(m+n)  (m->len(pat),n->len(txt))
# Space Complexity: O(m)

File: KMP.java
public class KMP {
    public static void constructLPS(String pat, int m, int[] lps) {
        int len = 0, i = 1;
        lps[0] = 0;
        while (i < m) {
            if (pat.charAt(i) == pat.charAt(len)) {
                lps[i++] = ++len;
            } else {
                if (len == 0)
                    lps[i++] = 0;
                else
                    len = lps[len - 1];
            }
        }
    }

    public static void KMPSearch(String pat, String txt) {
        int m = pat.length(), n = txt.length();
        int[] lps = new int[m];
        constructLPS(pat, m, lps);
        int i = 0, j = 0;
        while (i < n) {
            if (txt.charAt(i) == pat.charAt(j)) {
                i++;
                j++;
                if (j == m) {
                    System.out.println("Pattern found at: " + (i - m));
                    j = lps[j - 1];
                }
            } else {
                if (j == 0) {
                    i++;
                } else {
                    j = lps[j - 1];
                }
            }
        }
    }

    public static void main(String args[]) {
        String txt = "aabaacaadaabaaba";
        String pat = "aaba";
        KMPSearch(pat, txt);
    }
}
